onnxscript/rewriter/generic_pattern.py

```diff
@@ -294,6 +294,78 @@ def make_node(
         return proto
 
 
+class PatternMatchResult:
+    """Stores information about a match if a match was successful."""
+
+    def __init__(
+        self,
+        pattern: "GenericPattern",
+        model_nodes: typing.Sequence[oir.Node],
+        pattern_nodes: typing.Sequence[oir.Node],
+        pattern_input_names: typing.Sequence[str],
+        pattern_output_names: typing.Sequence[str],
+    ):
+        assert len(model_nodes) == len(pattern_nodes), (
+            f"The number of matched nodes in the model {len(model_nodes)} "
+            f"should the same as the number of nodes in the pattern "
+            f"{len(pattern_nodes)}."
+        )
+        self.pattern = pattern
+        self.model_nodes = model_nodes
+        self.pattern_nodes = pattern_nodes
+        self.pattern_input_names = pattern_input_names
+        self.pattern_output_names = pattern_output_names
+        self.kwargs = {}
+
+        matched_pattern_to_model_name: dict[str, str] = {}
+        for gn, pn in zip(model_nodes, pattern_nodes):
+            assert (
+                gn.op_type == pn.op_type
+            ), f"Unexpected type mismatch {gn.op_type!r} != {pn.op_type!r}"
+            assert len(gn.input_names) == len(
+                pn.input_names
+            ), f"Unexpected number of inputs for type {gn.op_type}"
+            for a, b in zip(gn.input_names, pn.input_names):
+                if b == "":
+                    # optional input or not an interesting input
```

Comment:
This could end up being a problematic convention: users may want to check that an optional input has the "" value sometimes and sometimes they may want to ignore an uninteresting input. It is better to fix the meaning of "" in a pattern-node-input: I think it is better to treat it as a requirement that the matching graph-node-input must also be "". (The users can always use an extra input symbol for the dont-care situation.)

Reply:
Created issue #1385